src/client.rs

```diff
@@ -77,6 +77,173 @@ pub extern "C" fn rustls_client_config_builder_build(
     }
 }
 
+/// Currently just a placeholder with no accessors yet.
+/// https://docs.rs/rustls/0.19.0/rustls/struct.RootCertStore.html
+#[allow(non_camel_case_types)]
+pub struct rustls_root_cert_store {
+    _private: [u8; 0],
+}
+
+/// A representation of the rustls Certificate type, which is an array of
+/// bytes, nominally in DER-encoded X.509.
+/// https://docs.rs/rustls/0.19.0/rustls/struct.Certificate.html
+#[allow(non_camel_case_types)]
+#[repr(C)]
+pub struct rustls_certificate {
+    bytes: *const u8,
+    len: usize,
+}
+
+/// Input to a custom certificate verifier callback. See
+/// rustls_client_config_builder_dangerous_set_certificate_verifier().
+#[allow(non_camel_case_types)]
+#[repr(C)]
+pub struct rustls_verify_server_cert_params {
+    end_entity: rustls_certificate,
+    intermediates: *const rustls_certificate,
+    intermediates_len: usize,
+    roots: *const rustls_root_cert_store,
+    dns_name: *const c_char,
+    dns_name_len: usize,
+    ocsp_response: *const u8,
+    ocsp_response_len: usize,
+}
+
+/// User-provided input to a custom certificate verifier callback. See
+/// rustls_client_config_builder_dangerous_set_certificate_verifier().
+#[allow(non_camel_case_types)]
+type rustls_verify_server_cert_user_data = *mut libc::c_void;
+
+// According to the nomicon https://doc.rust-lang.org/nomicon/ffi.html#the-nullable-pointer-optimization):
+// > Option<extern "C" fn(c_int) -> c_int> is a correct way to represent a
+// > nullable function pointer using the C ABI (corresponding to the C type int (*)(int)).
+// So we use Option<...> here. This is the type that is passed from C code.
+#[allow(non_camel_case_types)]
+type rustls_verify_server_cert_callback = Option<
+    unsafe extern "C" fn(
+        userdata: rustls_verify_server_cert_user_data,
+        params: *const rustls_verify_server_cert_params,
+    ) -> rustls_result,
```

Comment:
Can this be `Option<NonNullVerifyCallback>` to avoid the duplication of the inner definition? Also I think `NonNullVerifyCallback` could just be `VerifyCallback`--the fact that it's non-null is implicit because it's not possible to have a value of `NonNullVerifyCallback` that isn't, well, a value. But I see the merit in disambiguating between the Rust callback type and the nullable type exposed to C so whatever you prefer here is fine by me.

Reply:
That's a shame about `cbindgen`. These definitions are part of the stable API and aren't likely to change so the duplication isn't so bad.